### PR TITLE
Revisited how transitions are handled by the cmx 3600 adapter.

### LINKED ISSUE #912: cmx_3600 adapter incorrectly processing three-clip transitions
I have a case of a three part Transition in a CMX 3600 EDL. It looks like so:

Clip A -> Dissolve -> Clip B -> Dissolve -> Clip C

Here's the EDL:
```
TITLE:   TRANSITION_TEST_2
FCM: NON-DROP FRAME
001  ABC0000. V     C        01:00:06:18 01:00:08:00 01:04:11:17 01:04:12:23
002  ABC0010. V     C        01:00:06:15 01:00:08:18 01:04:12:23 01:04:15:02
003  ABC0010. V     C        01:00:08:18 01:00:08:18 01:04:15:02 01:04:15:02
003  ABC0020. V     D    035 01:00:06:22 01:00:10:07 01:04:15:02 01:04:18:11
* BLEND, DISSOLVE
004  ABC0020. V     C        01:00:10:07 01:00:10:07 01:04:18:11 01:04:18:11
004  ABC0030. V     D    064 01:00:06:10 01:00:09:22 01:04:18:11 01:04:21:23
* BLEND, DISSOLVE
005  ABC0040. V     C        01:00:08:14 01:00:12:14 01:04:21:23 01:04:25:23
```

In that example, ABC0020 should have a duration of 97, however OTIO it showing it to have 50 frames.

There appears to be something specifically wrong here, in the logic that derives what is the "previous clip":
https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/84af5e5a30e4d14cf3434d6d055631e115ac1ee5/src/py-opentimelineio/opentimelineio/adapters/cmx_3600.py#L706-L717

### LINKED ISSUE #977: EDL Adapter incorrectly processing EDL that ends with "hanging dissolve event"
I have an EDL that ends with a Transition between two clips.

In the EDL, it looks like this, a "hanging dissolve event":

```
TITLE:   TRANSITION_TEST_2
FCM: NON-DROP FRAME
001  ABC0000. V     C        01:00:06:18 01:00:08:00 01:04:11:17 01:04:12:23
002  ABC0010. V     C        01:00:06:15 01:00:08:18 01:04:12:23 01:04:15:02
003  ABC0010. V     C        01:00:08:18 01:00:08:18 01:04:15:02 01:04:15:02
003  ABC0020. V     D    035 01:00:06:22 01:00:10:07 01:04:15:02 01:04:18:11
* BLEND, DISSOLVE
```

That should result in an OTIO that includes three Clips with a Transition between the last two clips, like this:

Screen Shot 2021-05-27 at 11 38 44 AM

However, the EDL Adapter is create a Transition to a Gap, like this:

Screen Shot 2021-05-26 at 10 50 31 AM

Test EDL attached.

Unittest:
```
    def test_hanging_transition(self):
        """
        Test A->B Transition with no ending clip.
        """
        tl = otio.adapters.read_from_file("dissolve_test_5.edl")
        self.assertEqual(len(tl.tracks[0]), 4)

        self.assertIsInstance(tl.tracks[0][0], otio.schema.Clip)
        self.assertIsInstance(tl.tracks[0][1], otio.schema.Clip)
        self.assertIsInstance(tl.tracks[0][2], otio.schema.Transition)

        self.assertEqual(tl.tracks[0][2].duration().value, 35.0)
        self.assertEqual(tl.tracks[0][3].duration().value, 96.0)

        self.assertIsInstance(tl.tracks[0][3], otio.schema.Clip)
```

### LINKED ISSUE #1328: CMX 3600 does not handle correctly transitions
### Incorrect Functionality and General Questions

CMX 3600 reads everything in an EDL as Clips and then perform a conversion to `Transitions` when needed. However, instead of keeping the Clip before the transition (in the EDL), it removes it from the `Timeline` and keeps the `Clip` for the EDL transition, loosing meta data and possibly media for `Clip` at the end of the transition (in the timeline). 

## To Reproduce
Load the following EDL:

```
TITLE:   OTIO_TEST
FCM: NON-DROP FRAME

001  ABC0100 V     C        01:00:00:00 01:00:01:00 01:00:00:00 01:00:01:00
* FROM CLIP NAME: shot_001_v001
* COMMENT: shot_001

002  ABC0200 V     C        01:00:01:00 01:00:01:00 01:00:01:00 01:00:01:00
* COMMENT: shot_002

002  ABC0200 V     D    024 01:00:01:00 01:00:02:00 01:00:01:00 01:00:02:00
* FROM CLIP NAME: shot_001_v001
* TO CLIP NAME: shot_002_v001

003  ABC0300 V     C        00:00:05:00 00:00:15:00 01:00:02:00 01:00:12:00
* FROM CLIP NAME: shot_003_v001
* COMMENT: shot_003
```
_note: The last entry is needed because of another problem with the cmx 3600 adapter when the last entry is a transition._

`edl_timeline = otio.adapters.read_from_string(edl_text, adapter_name="cmx_3600")`

Check the result by looping over children:
```
for child in edl_timeline.tracks[0]:
    logger.info("Child %s" % child)
```

This gives:

```
Clip("shot_001_v001", MissingReference('', None, {}), TimeRange(RationalTime(86400, 24), RationalTime(36, 24)), {'cmx_3600': {'comments': ['COMMENT: shot_001'], 'reel': 'ABC0100'}})
Transition("shot_002_v001", "SMPTE_Dissolve", RationalTime(12, 24), RationalTime(12, 24), {'cmx_3600': {'reel': 'ABC0200', 'transition': 'D', 'transition_duration': 24.0}})
Clip("shot_002_v001", MissingReference('', None, {}), TimeRange(RationalTime(86436, 24), RationalTime(12, 24)), {'cmx_3600': {'reel': 'ABC0200', 'transition': 'D', 'transition_duration': 24.0}})
Clip("shot_003_v001", MissingReference('', None, {}), TimeRange(RationalTime(120, 24), RationalTime(240, 24)), {'cmx_3600': {'comments': ['COMMENT: shot_003'], 'reel': 'ABC0300'}})
```

The `Clip` after the `Transition` has wrong meta data, since it is the `Clip` which was created when reading the EDL, before treating transitions. The same would apply for media reference if there was some.

## Expected Behavior

The `Transition` should be inserted between the `Clips` contributing to the transition, and those `Clips` should be kept, with just their time range adjusted.

### PR DESCRIPTION
**Pull request by gplsteph**
The code handling transitions in the cmx 3600 adapter was hard to follow and was introducing problems. These changes
fix these problems and make the code easier to maintain.

- Transitions lines are now collected when parsing EDL lines and treated by the `ClipHandler`.
- The `Clip` source and record points are replaced with the transition values when the line is parsed.
- The transition `in_offset` is now 0 and the `out_offset` is the transition duration, matching in the timeline what would be done if the editing was read from another format.
- The transition is added just before the clip when it is added to a track, and that's it.
- Adjusted tests for the new way of handling transitions, added some other checks.
- Added an EDL read/write/read roundtrip to check that the reader and writer are compatible.
- Removed code which was not needed anymore.

```Fixes #1328```
```Fixes #912```
```Fixes #977```

Now the transitions read from an EDL appear at the right time in the timeline, and the timeline has the right overall length.
EDL transitions

**Associated tests.**
These changes are covered by the `test_cmx_3600_adapter` tests. A read/write/read roundtrip test was added to ensure they are compatible with the writer.

### CODE REVIEW

**Comment by jminor** [comment]
The failed jobs look like a temporary network problem. I pressed the "Re-run failed jobs" button.

**Comment by gplsteph** [comment]
Hello, I signed the CLA, but the PR still says "uncovered". Should I push an empty commit to get the bot ran again and have it updated?

**Comment by jminor** [comment]
Hmm.. I would have expected it to update automatically. Can you try resolving the DCO check first? Maybe that will re-run the CLA check also. We only recently enabled EasyCLA, so we're still learning it's quirks.
